### PR TITLE
get_config_directory: Use `files` origin

### PR DESCRIPTION
**Pull request by sils1297**
Fixes https://github.com/coala-analyzer/coala/issues/2030

### CODE REVIEW

File: tests/settings/ConfigurationGatheringTest.py [hunk 1]
```diff
@@ -285,39 +282,6 @@ def test_no_config(self):
...
-        os.path.isdir = lambda *args: True
-        config_dir = get_config_directory(section)
-        self.assertEqual(config_dir, "/path/to/dir/config")
-
-        os.path.isdir = lambda *args: False
-        config_dir = get_config_directory(section)
-        self.assertEqual(config_dir, "/path/to/dir")
-
-        os.path.isdir = old_isdir
-        os.path.isfile = old_isfile
```
Comment (Makman2) [inline tests/settings/ConfigurationGatheringTest.py]: at least one test case should test `get_config_directory` independently^^

File: coalib/settings/ConfigurationGathering.py [hunk 2]
```diff
@@ -200,15 +200,46 @@ def find_user_config(file_path, max_trials=10):
...
+    True
+
+    If the ``files`` setting is given with an originating coafile, the directory
+    of the coafile will be assumed the configuration directory:
+
+    >>> section.append(Setting("files", "**", origin="/tmp/.coafile"))
+    >>> get_config_directory(section) == os.path.abspath('/tmp')
+    True
+
+    However if it's origin is already a directory this will be preserved:
```
Comment (Makman2) [inline coalib/settings/ConfigurationGathering.py]: it's -> its

File: coalib/settings/ConfigurationGathering.py [hunk 3]
```diff
@@ -200,15 +200,46 @@ def find_user_config(file_path, max_trials=10):
...
+
+    Given an empty section:
+
+    >>> section = Section("name")
+
+    The configuration directory is not defined and will therefore fallback to
+    the current directory:
+
+    >>> dir = get_config_directory(section)
+    >>> dir == os.path.abspath(".")
```
Comment (Makman2) [inline coalib/settings/ConfigurationGathering.py]: make a oneliner^^

File: tests/settings/ConfigurationGatheringTest.py (hunk 1 above)
Comment (sils1297) [inline tests/settings/ConfigurationGatheringTest.py]: there's doctests they cover all valid cases with logical explanations

File: coalib/settings/ConfigurationGathering.py [hunk 4]
```diff
@@ -200,15 +200,47 @@ def find_user_config(file_path, max_trials=10):
...
+    of the coafile will be assumed the configuration directory:
+
+    >>> section.append(Setting("files", "**", origin="/tmp/.coafile"))
+    >>> get_config_directory(section) == os.path.abspath('/tmp/')
+    True
+
+    However if its origin is already a directory this will be preserved (the
+    path mangling is needed for the example to work on windows as well):
+
+    >>> tmpdir = os.path.dirname(os.path.abspath("/tmp")+ "/")
```
Comment (AbdealiJK) [inline coalib/settings/ConfigurationGathering.py]: Newcomer issue: space before `+`

Comment (AbdealiJK) [inline coalib/settings/ConfigurationGathering.py]: Could you print tmpdir here ? So that the reader can understand what it is better ? Maybe even do `os.isdir(tmpdir)` to stress your point ?

Comment (AbdealiJK) [inline coalib/settings/ConfigurationGathering.py]: Actually dont do isdir ... it wouldnt work in windows.
Just print `tmpdir` ?

File: tests/coalaCITest.py [hunk 5]
```diff
@@ -105,6 +105,9 @@ def test_coala_delete_orig(self):
              make_temp(suffix='.coafile', prefix='', dir=tempdir) as coafile,\
              make_temp(dir=tempdir) as unrelated_file:
             orig_file.close()
+            with open(coafile, mode="w") as file:
+                file.write("files = something\n")
+
```
Comment (AbdealiJK) [inline tests/coalaCITest.py]: Why is this here ? Is it needed for `test_coala_delete_orig` ? That's confusing don't you think ?

Comment (sils1297) [inline tests/coalaCITest.py]: it's needed as the new function looks for where the files setting is defined, not which config file is used (so if the files setting comes from CLI the cwd is the project dir and not the dir of the config file which might be somewhere else)

File: coalib/settings/ConfigurationGathering.py (hunk 4 above)
Comment (sils1297) [inline coalib/settings/ConfigurationGathering.py]: if I print it the output will differ across platforms, that's the issue :( I have no idea how to fix this here...

**Comment by SanketDG** [comment]
ack af03e26

**Comment by SanketDG** [comment]
ack d4aa35e aabc7c7

**Comment by sils1297** [comment]
@rultor merge

**Comment by sils1297** [comment]
@SanketDG thanks! :)

**Comment by sils1297** [comment]
ack c697c52

**Comment by SanketDG** [comment]
@rultor merge
